Ask the client before pursuing an immersive request

jeanhbr opened the pull request:
#### 6264e5e9e22e06ecb92f44e11892713bbce07a09
<pre>
Ask the client before pursuing an immersive request
<a href="https://bugs.webkit.org/show_bug.cgi?id=303145">https://bugs.webkit.org/show_bug.cgi?id=303145</a>
<a href="https://rdar.apple.com/165447324">rdar://165447324</a>

Reviewed by Etienne Segonzac.

Add a WK delegate protocol to communicate with the client and ask consent
before pursuing an immersive request.

Tests: model-element/immersive/model-element-immersive-basic.html
       model-element/immersive/model-element-immersive-no-client-approval.html

* LayoutTests/model-element/immersive/model-element-immersive-basic-expected.txt: Added.
* LayoutTests/model-element/immersive/model-element-immersive-basic.html: Renamed from LayoutTests/model-element/model-element-immersive.html.
* LayoutTests/model-element/immersive/model-element-immersive-no-client-approval-expected.txt: Added.
* LayoutTests/model-element/immersive/model-element-immersive-no-client-approval.html: Added.
* LayoutTests/model-element/model-element-immersive-expected.txt: Removed.
Move the immersive-related model element tests inside a subfolder,
and add a test that verifies that an immersive request fails if the client doesn&apos;t allow it.

* Source/WebCore/dom/DocumentImmersive.cpp:
(WebCore::DocumentImmersive::requestImmersive):
Add checks before pursuing an immersive request.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::canEnterImmersiveElement const):
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _immersiveEnvironmentDelegate]):
(-[WKWebView _setImmersiveEnvironmentDelegate:]):
(-[WKWebView _canEnterImmersiveElementFromURL:completion:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKImmersiveEnvironmentDelegate.h: Added.
New delegate protocol to communicate with the client for immersive-related actions.

* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::canEnterImmersiveElementFromURL):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::canEnterImmersiveElementFromURL):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::canEnterImmersiveElementFromURL):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::canEnterImmersiveElement const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::canEnterImmersiveElement):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
Plumbing the request up to the UI Process.

* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::shouldAcceptImmersiveEnvironmentRequests const):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView initWithFrame:configuration:]):
(-[TestRunnerWKWebView webView:canPresentImmersiveEnvironmentFromURL:completion:]):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::TestController::platformResetStateToConsistentValues):
Add a test option that forces the test runner to either allow or disallow immersive presentations.

Canonical link: <a href="https://commits.webkit.org/303611@main">https://commits.webkit.org/303611@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e62c8445e4f0883cf15523b5b16c51837dc6cbb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132993 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5494 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44089 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140529 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134863 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5926 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5357 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101699 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135939 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119166 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82498 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1660 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83763 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/113100 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37284 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143182 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5163 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37865 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110077 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5245 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4421 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110258 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27950 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3962 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115428 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/58734 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5217 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/33780 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5058 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68669 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5307 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/5175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

etiennesegonzac commented on Source/WebCore/dom/DocumentImmersive.cpp:
```diff
@@ -81,9 +81,31 @@ void DocumentImmersive::exitImmersive(Document& document, RefPtr<DeferredPromise
+    enum class EmitErrorEvent : bool { No, Yes };
+    auto handleError = [weakElement = WeakPtr { *element }, weakThis = WeakPtr { *this }](ASCIILiteral message, EmitErrorEvent emitErrorEvent, CompletionHandler<void(ExceptionOr<void>)>&& completionHandler) mutable {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis)
```
and weakElement? (it's deref'ed below)

etiennesegonzac commented on Source/WebCore/dom/DocumentImmersive.cpp:
```diff
@@ -81,9 +81,31 @@ void DocumentImmersive::exitImmersive(Document& document, RefPtr<DeferredPromise
+    if (RefPtr window = document().window(); !window || !window->consumeTransientActivation())
+        return handleError("Cannot request immersive without transient activation."_s, EmitErrorEvent::Yes, WTFMove(completionHandler));
+
+    page()->chrome().client().canEnterImmersiveElement(*element, [handleError, completionHandler = WTFMove(completionHandler)](auto canEnterImmersive) mutable {
```
page() could be null too

etiennesegonzac replied:
maybe we can use document().protectedPage() directly (and nullcheck)

etiennesegonzac commented on Source/WebKit/UIProcess/API/Cocoa/_WKImmersiveEnvironmentDelegate.h:
```diff
@@ -0,0 +1,39 @@
+WK_SWIFT_UI_ACTOR
+WK_API_AVAILABLE(visionos(WK_XROS_TBA))
+@protocol _WKImmersiveEnvironmentDelegate <NSObject>
+
```
let's reference a FIXME with the issue tracking api review (even if we only do it once the delegate is complete)

etiennesegonzac commented on Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
```diff
@@ -95,6 +95,10 @@
 @property (nonatomic, setter=_setScrollingUpdatesDisabledForTesting:) BOOL _scrollingUpdatesDisabledForTesting;
 
+#if ENABLE(MODEL_ELEMENT_IMMERSIVE)
+@property (nonatomic, assign) BOOL shouldPresentImmersiveEnvironments;
```
maybe `shouldAcceptImmersiveEnvironmentRequests`?

etiennesegonzac commented on LayoutTests/model-element/immersive/model-element-immersive-no-client-approval.html:
```diff
@@ -0,0 +1,29 @@
+<!DOCTYPE html> <!-- webkit-test-runner [ ModelElementEnabled=true ModelProcessEnabled=true ModelElementImmersiveEnabled=true ] -->
```
for readability maybe we should add the option set to false here

etiennesegonzac reviewed (commented):
Small type in the commit message: "an immersive request**s**".

jeanhbr committed:
Ask the client before pursuing an immersive request
https://bugs.webkit.org/show_bug.cgi?id=303145
rdar://165447324

Reviewed by Etienne Segonzac.

Add a WK delegate protocol to communicate with the client and ask consent
before pursuing an immersive request.

Tests: model-element/immersive/model-element-immersive-basic.html
       model-element/immersive/model-element-immersive-no-client-approval.html

* LayoutTests/model-element/immersive/model-element-immersive-basic-expected.txt: Added.
* LayoutTests/model-element/immersive/model-element-immersive-basic.html: Renamed from LayoutTests/model-element/model-element-immersive.html.
* LayoutTests/model-element/immersive/model-element-immersive-no-client-approval-expected.txt: Added.
* LayoutTests/model-element/immersive/model-element-immersive-no-client-approval.html: Added.
* LayoutTests/model-element/model-element-immersive-expected.txt: Removed.
Move the immersive-related model element tests inside a subfolder,
and add a test that verifies that an immersive request fails if the client doesn't allow it.

* Source/WebCore/dom/DocumentImmersive.cpp:
(WebCore::DocumentImmersive::requestImmersive):
Add checks before pursuing an immersive request.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::canEnterImmersiveElement const):
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _immersiveEnvironmentDelegate]):
(-[WKWebView _setImmersiveEnvironmentDelegate:]):
(-[WKWebView _canEnterImmersiveElementFromURL:completion:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKImmersiveEnvironmentDelegate.h: Added.
New delegate protocol to communicate with the client for immersive-related actions.

* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::canEnterImmersiveElementFromURL):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::canEnterImmersiveElementFromURL):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::canEnterImmersiveElementFromURL):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::canEnterImmersiveElement const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::canEnterImmersiveElement):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
Plumbing the request up to the UI Process.

* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::shouldAcceptImmersiveEnvironmentRequests const):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView initWithFrame:configuration:]):
(-[TestRunnerWKWebView webView:canPresentImmersiveEnvironmentFromURL:completion:]):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::TestController::platformResetStateToConsistentValues):
Add a test option that forces the test runner to either allow or disallow immersive presentations.

Canonical link: https://commits.webkit.org/303611@main
